forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,179 @@
+package forge.game.ability.effects;
+
+import com.google.common.base.Predicates;
+import com.google.common.collect.Iterables;
+import forge.game.Game;
+import forge.game.ability.AbilityKey;
+import forge.game.ability.AbilityUtils;
+import forge.game.ability.SpellAbilityEffect;
+import forge.game.card.Card;
+import forge.game.card.CardCollection;
+import forge.game.card.CardLists;
+import forge.game.card.CardZoneTable;
+import forge.game.cost.CostDiscard;
+import forge.game.cost.CostPart;
+import forge.game.cost.CostReveal;
+import forge.game.player.Player;
+import forge.game.player.PlayerController;
+import forge.game.spellability.LandAbility;
+import forge.game.spellability.SpellAbility;
+import forge.game.trigger.TriggerType;
+import forge.game.zone.PlayerZone;
+import forge.game.zone.Zone;
+import forge.game.zone.ZoneType;
+import forge.util.CardTranslation;
+import forge.util.Localizer;
+
+import java.util.List;
+import java.util.Map;
+
+public class DiscoverEffect extends SpellAbilityEffect {
+
+    @Override
+    protected String getStackDescription(SpellAbility sa) {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(sa.getActivatingPlayer()).append(" discovers ");
+        sb.append(sa.getParamOrDefault("Num", "1")).append(".");
+
+        return sb.toString();
+    }
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card host = sa.getHostCard();
+        final Game game = host.getGame();
+        final Player p = sa.getActivatingPlayer();
+
+        // Exile cards from the top of your library until you exile a nonland card with <N> mana value or less.
+        final String val = sa.getParamOrDefault("Num", "1");
+        final int num = AbilityUtils.calculateAmount(host, val, sa);
+
+        if (p == null || !p.isInGame()) return;
+
+        Card found = null;
+        CardCollection exiled = new CardCollection();
+        CardCollection rest = new CardCollection();
+
+        final PlayerZone library = p.getZone(ZoneType.Library);
+
+        for (int i = 0; i < library.size(); i++) {
+            final Card c = library.get(i);
+            exiled.add(c);
+            if (c.isValid("Card.nonLand+cmcLE" + val, sa.getActivatingPlayer(), host, sa)) {
```

Comment:
i don't know if i like the IsValid string.

when there is nothing that would be changed by text stuff, or is coming from different parameters,

wouldn't it better to just hard code the nonLand and cmc checks?